packages/server/src/adapters/content-handlers/selectContentHandlerOrUnsupportedMediaType.ts

```diff
@@ -1,18 +1,30 @@
 import { TRPCError } from '../../@trpc/server';
 import type { BaseContentTypeHandler } from '../../@trpc/server/http';
+// eslint-disable-next-line no-restricted-imports
+import type { Maybe } from '../../unstable-core-do-not-import/types';
 
 export function selectContentHandlerOrUnsupportedMediaType<
   THandlerOpts,
   THandler extends BaseContentTypeHandler<THandlerOpts>,
 >(handlers: THandler[], opts: THandlerOpts) {
-  const handler = handlers.find((handler) => handler.isMatch(opts));
+  let handler: THandler | undefined;
+  let received: Maybe<string> = undefined;
+
+  for (const h of handlers) {
+    const match = h.isMatch(opts);
+    received = match.received;
+    if (match.match) {
+      handler = h;
+      break;
+    }
+  }
```

Comment:
As discussed in discord, I think best to do this in a way which is fairly linear. iterating over and taking a match object from the last in the chain seems to create complexity

We probably want to already know the header at this level in the stack and just ask the content handler if it can handle that type?